lxd/storage_zfs.go

```diff
@@ -625,22 +625,31 @@ func (s *storageZfs) ImageCreate(fingerprint string) error {
 		return err
 	}
 
+	cleanup := func(err error) error {
+		if zerr := s.zfsDestroy(fs); zerr != nil {
+			err = fmt.Errorf("%s  During cleanup: %s", err, zerr)
+		}
+		if shared.PathExists(subvol) {
+			if oserr := os.Remove(subvol); oserr != nil {
```

Comment:
On Wed, Jul 20, 2016 at 11:16:19AM -0700, Sean Christopherson wrote:
> > @@ -625,22 +625,31 @@ func (s *storageZfs) ImageCreate(fingerprint string) error {
> >  		return err
> >  	}
> >  
> > +	cleanup := func(err error) error {
> > +		if zerr := s.zfsDestroy(fs); zerr != nil {
> > +			err = fmt.Errorf("%s  During cleanup: %s", err, zerr)
> > +		}
> > +		if shared.PathExists(subvol) {
> > +			if oserr := os.Remove(subvol); oserr != nil {
> 
> I don't think so?  I'm pretty sure the only way subvol would be non-empty is if zfsDestroy failed, in which case I think both Remove and RemoveAll would fail due to "resource busy".
> 
> Hold off on merging this for now, I'll follow-up with the submitter of #2194 to find out what exactly was left laying around after the failed image pull.

But if the directory is empty, then I don't see how this fixes the
problem in #2194; isn't that about leaving stuff around that actually
takes up space?

> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/lxc/lxd/pull/2222/files/16015f0ab699178f206e4cc9e2c4a4b6e182d976#r71578114